course3/homework-string/my_string.cc

```diff
@@ -0,0 +1,108 @@
+#include "my_string.h"
+
+#include <cstring>
+
+
+MyString::MyString(const char* str)
+{
+	if (str[0] != '\0')
+	{
+		mBuffer.reset(new char[std::strlen(str) + 1]);
+		std::strcpy(mBuffer.get(), str);
+	}
+}
+
+MyString::MyString(const MyString& str)
+{
+	if (!str.empty())
+	{
+		mBuffer.reset(new char[str.size() + 1]);
+		std::strcpy(mBuffer.get(), str.mBuffer.get());
+	}
+}
+
+bool MyString::empty() const
+{
+	return !mBuffer;
+}
+
+std::size_t MyString::size() const
+{
+	return empty() ? 0 : std::strlen(mBuffer.get());
+}
+
+const char* MyString::c_str() const
+{
+	return empty() ? "" : mBuffer.get();
+}
+
+MyString& MyString::operator=(const char* str)
+{
+	if (str[0] == '\0')
+	{
+		mBuffer.reset(nullptr);
+	}
+	else
+	{
+		mBuffer.reset(new char[std::strlen(str) + 1]);
+		std::strcpy(mBuffer.get(), str);
+	}
+	
+	return *this;
+}
+
+MyString& MyString::operator=(const MyString& str)
+{
+	if (str.empty())
+	{
+		mBuffer.reset(nullptr);
+	}
+	else
+	{
+		mBuffer.reset(new char[str.size() + 1]);
+		std::strcpy(mBuffer.get(), str.mBuffer.get());
+	}
+	
+	return *this;
+}
+
+bool MyString::operator==(const MyString& str) const
+{
+	if (empty() && str.empty()) return true;
+	if (empty() || str.empty()) return false;
+
+	return std::strcmp(mBuffer.get(), str.mBuffer.get()) == 0;
+}
+
+bool MyString::operator!=(const MyString& str) const
+{
+	return !operator==(str);
+}
+
+const char& MyString::operator[](int idx) const
+{
+	return mBuffer.get()[idx];
+}
+
+char& MyString::operator[](int idx)
+{
+	return mBuffer.get()[idx];
+}
+
+MyString MyString::operator+(const MyString& str) const
+{
+	if (empty()) return MyString(str);
+	else if (str.empty()) return MyString(*this);
```

Comment:
Is it okay to just return the object directly instead of invoking the copy constructor (i.e. will the return be copied by value anyway)?

In short, can I instead do this?:
if (empty()) return str;
else if (str.empty()) return *this;